ement.el

```diff
@@ -141,10 +139,6 @@ session and log in again."
         (let ((ement-save-session nil))
           (call-interactively #'ement-connect)))
     ;; Log in to new session.
-    (unless (string-match (rx bos "@" (group (1+ (not (any ":")))) ; Username
-                              ":" (group (optional (1+ (not (any blank)))))) ; Server name
-                          user-id)
-      (user-error "Invalid user ID format: use @USERNAME:SERVER"))
     (let* ((username (match-string 1 user-id))
```

Comment:
Since you've removed the `string-match` and are returning a plist instead, you need to bind the variables to the returned plist values, no?  Why rely on the global match data across function calls when you're doing that in a separate function?  (Please use `pcase-let*`).

Reply:
I'd gladly bind the parsed values directly but that requires the interface of `ement-connect` to change, e.g. its arglist would become `(&key user host password)` or just `(&rest properties)`.  Which one you'd prefer?

Reply:
I'd prefer to keep the interface as it is.  Is it really necessary to return a plist here?

Reply:
We're reading both user-id and password in the `interactive` form.  user-id needs to be split to do a password lookup.  As we read password in the `interactive` form, it has to be split in the `interactive` form.  Then it has to get split the same way the second time in the function body because bindings established within `interactive` form are not avaliable in the body (unless we use a let over defun).  I remember trying to capture the results of in-`interactive` computations once.  `&aux` didn't help, so let over defun was the only way to go.  Or we could simply reuse the saved match-data which I did initially.

Otherwise, we could simply repeat the splitting proceduce in the function body.  This means repeating the same computation twice.  If you're fine with that, I'll do it.  I myself would never do it if another solution would be readily available: if for any reason two results diverge, it will be a baffling bug, and any side-effects in the splitting procedure, or in other future in-`interactive` computations would become unsafe.

Reply:
I think it's more important to keep the command's signature sensible (from a user's point of view) than to avoid calling `string-match` one extra time in a function that is called once per session.  I don't think the MXID-splitting regexp is going to change of its own accord, but it could go into a variable, or the splitting itself could be done in a function.

Reply:
OK, I'll re-split in the `let*` preserving the function signature.  Will re-submit tomorrow, likely.  I think I understand the way you want it good enough.